src/DB/DumpGenerator.php

```diff
@@ -54,39 +87,137 @@ class DumpGenerator
+    {
+        if (empty($databases)) {
+            $connections = array_values(array_intersect_key(
+                array_flip(self::CONNECTION_MAP),
+                $this->dbConfig->get()[DbConfig::KEY_CONNECTION] ?? []
+            ));
+        } else {
+            if (!$this->validateDatabaseNames($databases)) {
+                return;
+            }
+            $connections = array_keys(array_intersect(self::DATABASE_MAP, $databases));
+            if (!$this->checkConnectionsAvailability($connections)) {
+                return;
+            }
+        }
+
+        foreach ($connections as $connection) {
+            $this->process(
+                self::DATABASE_MAP[$connection],
+                $this->connectionFactory->create($connection),
+                $removeDefiners
+            );
+        }
+    }
+
+    /**
+     * Validates database names
+     *
+     * @param array $databases
+     * @return bool
+     */
+    private function validateDatabaseNames(array $databases): bool
+    {
+        $invalidDatabaseNames = array_diff($databases, self::DATABASE_MAP);
+        if (!empty($invalidDatabaseNames)) {
+            $this->logger->error(sprintf(
+                'Incorrect the database names:[ %s ]. Available database names: [ %s ]',
+                implode(' ', $invalidDatabaseNames),
+                implode(' ', self::DATABASE_MAP)
+            ));
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Checks availability of connections
+     *
+     * @param array $connections
+     * @return bool
+     * @throws ConfigException
+     */
+    private function checkConnectionsAvailability(array $connections): bool
+    {
+        $result = true;
+        $envConnections = $this->dbConfig->get()[DbConfig::KEY_CONNECTION] ?? [];
+        foreach ($connections as $connection) {
+            if (!isset($envConnections[self::CONNECTION_MAP[$connection]])) {
+                $this->logger->error(sprintf(
+                    'Environment has not connection `%s` associated with database `%s`',
+                    self::CONNECTION_MAP[$connection],
+                    self::DATABASE_MAP[$connection]
+                ));
+                $result = false;
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * Creates database dump and archives it for each database.
      *
      * Lock file is created at the beginning of dumping.
      * This file has dual purpose, it creates a lock, so another DB backup process cannot be executed,
      * as well as serves a log with the name of created dump file.
      * If any error happened during dumping, dump file is removed.
      *
+     * @param string $database
+     * @param ConnectionInterface $connectionData
      * @param bool $removeDefiners
-     * @return void
-     * @throws \Magento\MagentoCloud\Package\UndefinedPackageException
+     * @throws UndefinedPackageException
      */
```

Comment:
ShellException can be thrown as well after you removed the catch statement.

Reply:
Added